### ISSUE TITLE
Using Gradle to build a jar does not include dependencies

### ISSUE THREAD

**Issue by jkpr** [opendatakit/javarosa#463]
#### Software versions 
JavaRosa v2.14.0, Java v11.0.1, operating system MacOS 10.13.6

#### Problem description
Using gradle to build a jar does not include dependencies.

#### Steps to reproduce the problem
Run `./gradlew jar` then inspect the contents of the jar after unzipping.

#### Expected behavior
Should include dependencies such as `joda-time`.

#### Suggested solution
Include this in the `jar{...}` block of `build.gradle`

```
from {configurations.compile.collect { it.isDirectory() ? it : zipTree(it) } }
```

This line exists in Validate. See: https://github.com/opendatakit/validate/blob/master/build.gradle#L86

**Comment by batkinson** [comment]
TLDR; This isn't necessary, or good practice, for the primary build. I request that we protect that primary and add a secondary distribution with deps if this is common enough to warrant catering to directly. 

Please, reconsider this train of thought. If every library did this gradle wouldn't have the ability to attempt to manage dependencies at the build level at all. They would all be bundled in every binary an unused ones extricated by binary analysis, which is imperfect and error-prone.

Said another way, dependency management systems like gradle's (actually Apache Maven's) are meant to allow *modular* libraries specifically to allow for aligning and resolving conflicts in downstream (consuming) projects. By bundling dependencies you make it impossible to use the library as a module, and you force downstream code to leverage imprecise blunt force tools like proguard and the like just to do what the dependency management tools were designed for in the first place (often with unexpected side-effects until you find all the dynamic references and explicitly configure keep rules).

If having another, secondary, distribution that bundles dependencies is helpful, provide that as a convenience, but please don't pollute the primary. It's backwards, especially since it's so easy to slurp all deps in with a clean (and accurate) dependency graph. That isn't possible when library x is also library y and z in disguise.

You actually provided a perfect case study with validate. Validate is not a good model for how to publish libraries. I suspect that this is probably due to the fact that it was _not really meant to be a library_. From https://github.com/opendatakit/validate:

> ODK Validate _is a desktop Java application_...

Javarosa is different. It was never meant to be a standalone app. 

Coincidentally, I recently had to author a validate alternative specifically to address these types of issues. I actually mention one specifically in the readme for the project: https://github.com/cims-bioko/validate.

**Comment by jkpr** [comment]
Those are good points. There must be a better way, then!

Here is the background. I modified Javarosa slightly to add a new XPath function. I created the jar, and added it to Validate with `compile fileTree(dir: 'libs', include: '*.jar')` inside the inside the `dependencies{ ... }` block after putting the jar in the `libs` dir of that project. I took out the `compile group: 'org.opendatakit', name: 'opendatakit-javarosa', version: '2.14.0'`.

Validate did not pull in `joda-time` (a dependency of Javarosa) when I built its jar. So my question is this: what is the proper way for Gradle to know to grab `joda-time` because it is needed for Javarosa in this case?

**Comment by yanokwa** [comment]
One option is to do what the [Gradle community recommends](https://discuss.gradle.org/t/how-to-include-dependencies-in-jar/19571) which is including the [Shadow plugin](https://github.com/johnrengelman/shadow). Then @j can call `./gradlew shadowJar`. Would that work for you @batkinson?

@batkinson Relatedly, I've been toying with the idea of adding CLI validation as a feature of JavaRosa (basically what you've done in CIMS Validate) as a first step in maybe getting rid of ODK Validate entirely.

**Comment by batkinson** [comment]
> One option is to do what the [Gradle community recommends](https://discuss.gradle.org/t/how-to-include-dependencies-in-jar/19571) which is including the [Shadow plugin](https://github.com/johnrengelman/shadow). Then @jkpr can call `./gradlew shadowJar`. Would that work for you @batkinson?

Yes, a shade task is even better. My own interest is simply to ensure jr is not unnecessarily coupled to the extent that it can't be leveraged as a simple library.

> @batkinson Relatedly, I've been toying with the idea of adding CLI validation as a feature of JavaRosa (basically what you've done in CIMS Validate) as a first step in maybe getting rid of ODK Validate entirely. It's mostly used by apps that have JavaRosa already and can call some validation function.

After seeing how odk validate was a thin wrapper around javarosa, I think that could make a lot of sense. I don't know if pushing end-user-style validation into javarosa is the direction I would go. I would probably want to keep any UI code separate and provide just enough framework for those tools to operate well (well-design, easily implemented, decoupled interfaces and commonly required stubs or stub builders for things that are difficult to provide when validating). In the end, that's essentially what odk validate did and it worked well. It just conflated things by trying to be a full-fledged app with a GUI too.

I think this is one of those rare areas in design where you can get the best of both. If you avoid coupling too much to the core library, you can build a CLI, a GUI, a web service, and other things maybe no one thought of. The cost is maintaining them as units and fighting the urge to conflate things. 

I used the library-ized validate to build [an elastic-scaling web api ](https://github.com/cims-bioko/validate-web) using dockerized validation service instances. Another application of that would be to use it to pull the core validation logic out of ODK Validate and focus it on providing an offline-capable graphical user experience. You could offer a CLI tool separately. Alternatively, you could just skip in-process validation altogether for those who have connectivity and use a web service similar to the one I just built to support more run-time environments (python apps using pyxform could drop needing to invoke a JVM process to validate, for example).

**Comment by batkinson** [comment]
> Here is the background. I modified Javarosa slightly to add a new XPath function. I created the jar, and added it to Validate with `compile fileTree(dir: 'libs', include: '*.jar')` inside the inside the `dependencies{ ... }` block after putting the jar in the `libs` dir of that project. I took out the `compile group: 'org.opendatakit', name: 'opendatakit-javarosa', version: '2.14.0'`.

If you are just adding a new xpath function, it should not normally require modifying and recompiling javarosa. You can add custom function handlers by adding them to the evaluation context, for example:

`formDef.getEvaluationContext().addFunctionHandler(new MyCustomFunctionHandler());`

> Validate did not pull in `joda-time` (a dependency of Javarosa) when I built its jar. So my question is this: what is the proper way for Gradle to know to grab `joda-time` when building Validate because it is needed for Javarosa?

When you removed the declarative compile dependency on javarosa, the tool can only look at your jars as simple files as opposed to libraries with dependencies themselves. With declarative dependencies, Gradle uses the metadata from Maven or Ivy artifact repositories to determine transitive dependencies, download them, and put them in the classpath during the build. By deciding to use jars, you (unknowningly?) decided to specify exactly what should be on the classpath for Gradle. There may be hacks to make what you're doing work, but I suspect you might be better off [learning more about declarative dependencies first](https://docs.gradle.org/current/userguide/declaring_dependencies.html).

I suspect there are better ways to achieve what you're attempting, but it's tough to say without having more detailed information (such as code).

**Comment by jkpr** [comment]
> If you are just adding a new xpath function, it should not normally require modifying and recompiling javarosa. You can add custom function handlers by adding them to the evaluation context, for example:

> formDef.getEvaluationContext().addFunctionHandler(new MyCustomFunctionHandler());

Thanks for that tip! I ran `grep` over Collect, and I saw that the `externalDataHandlerPull` is added this way.

As for which way to do it, via Javarosa or custom function handler, Mitch had a more nuanced opinion: https://forum.opendatakit.org/t/adding-custom-function-handlers/2145/2. He wrote that there are good reasons to modify Javarosa.

Thinking about what positive we can take away, should this be officially documented online anywhere?

**Comment by jkpr** [comment]
> When you removed the declarative compile dependency on javarosa, the tool can only look at your jars as simple files as opposed to libraries with dependencies themselves. With declarative dependencies, Gradle uses the metadata from Maven or Ivy artifact repositories to determine transitive dependencies, download them, and put them in the classpath during the build.

Thanks! I saw this (https://docs.gradle.org/current/userguide/declaring_dependencies.html) when searching about including dependencies. I didn't read closely enough!

As for how to get the dependencies without going through an online repository, such as in my situation I described earlier, perhaps it is the case that the Gradle way of doing things would be to use a multi-project build. I'm not sure how practical that is here since Javarosa and Validate are separate repositories on github, though.

**Comment by batkinson** [comment]
> As for which way to do it, via Javarosa or custom function handler, Mitch had a more nuanced opinion: https://forum.opendatakit.org/t/adding-custom-function-handlers/2145/2. He wrote that there are good reasons to modify Javarosa.

Mitch's statement does not conflict with what wrote. He is describing two approaches and when to use them. I still don't know what you are actually attempting and why, outside of adding an xpath function to javarosa and validate. Validate already registers custom handlers, so what you're doing would not be necessary unless you are trying to extend core functionality. If you are trying to change the entire ODK suite with new core functionality, adding it to jr would be consistent with existing code (though that run-on if/else chain is not good design in my opinion, fwiw). However, if you are looking to do this, why would you need to revamp the build? 

> Thinking about what positive we can take away, should this be officially documented online anywhere?

Good question. I don't think I understand your issue well enough to suggest anything of value myself. I think understanding what exactly you're doing is necessary to help any further. Maybe hop on the dev chat and talk to someone in real-time? What you're attempting seems a bit unusual. You might save yourself some time, and learn more about the various pieces involved at the same time.

**Comment by jkpr** [comment]
> I still don't know what you are actually attempting and why, outside of adding an xpath function to javarosa and validate.

Outside of those two things, we are trying to update Validate in `pyxform` and Javarosa in `ODK Collect`. This allows us to successfully validate our XLSForms and take advantage of the custom function in ODK Collect.

The problem that necessitated all of this was whitespace being added inadvertently to the end of text entry on ODK Collect. For example, our data collectors would enter `"John "` or `"John\n"` instead of `John` and that didn't work well with later calculations inside the questionnaire.

This will soon be moot thanks to https://github.com/opendatakit/javarosa/pull/467

> Validate already registers custom handlers

For me, it would have been great if I had known about registering custom handlers. That is what I meant about documenting for future, but I'm not certain where that would go and I do not know the best way to do it. I might have gone that route instead of forking Javarosa, Validate, pyxform, and ODK Collect.

I have seen some requests to add other calculate functions to ODK on the forum, but the community is especially hesitant if it is not in the XForm spec. Registering custom handlers is probably the best way to do that.

**Comment by batkinson** [comment]
> This will soon be moot thanks to #467

Ah, ok. If you referenced it earlier, it would have provided more of the context I needed to help.

> For me, it would have been great if I had known about registering custom handlers. 

As you saw from Mitch's old post, built-ins were not leveraging it much. I think the reality is that the pain involved in this type of extension means few choose to go this far. Also, documenting this type of thing basically advertises and, in a way, implicitly endorses it. The problem is that it requires forking nearly every project in the suite to make it work in concert like the rest. If you consider what that looks like in the aggregate, endorsing that doesn't end well.

> I have seen some requests to add other calculate functions to ODK on the forum, but the community is especially hesitant if it is not in the XForm spec. Registering custom handlers is probably the best way to do that.

I have been in your position. I think the reality is that the functions provided by ODK should be limited to the smallest set that provides the most power. Otherwise, it becomes unreasonable for developers as well as users. This is one very important reason to favor the xpath/xforms specs. If there was a lightweight way to extend without promoting forks and without polluting or conflicting with the standard namespace it would have little down-side in my opinion. If only you could register custom functions using a namespace, defined in a form using a scripting language, then you could even validate with them. :)

**Comment by lognaturel** [comment]
Thanks for the good discussion, here. I think we now agree that there is no immediate action needed so I will close.
